source/common/upstream/cluster_manager_impl.h

```diff
@@ -327,6 +328,42 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     bool destroying_{};
   };
 
+  class GrpcMuxFactoryImpl : public Config::GrpcMuxFactory {
+  public:
+    GrpcMuxFactoryImpl() : muxes_() {}
+
+    Config::GrpcMux* getOrCreateMux(const LocalInfo::LocalInfo& local_info, Grpc::AsyncClientPtr async_client,
+                                    Event::Dispatcher& dispatcher, const Protobuf::MethodDescriptor& service_method,
+                                    Runtime::RandomGenerator& random,
+                                    const ::envoy::api::v2::core::ApiConfigSource& config_source,
+                                    Stats::Scope& scope, const std::string type_url) override {
+
+      std::string mux_key; // config_source_name + type_url
+      switch(config_source.grpc_services(0).target_specifier_case()) {
+        case envoy::api::v2::core::GrpcService::kEnvoyGrpc: {
+          mux_key = config_source.grpc_services(0).envoy_grpc().cluster_name() + type_url;
```

Comment:
I wonder if we should literally require proto equality here. The reasoning here is that there are configuration options, e.g. timeouts, initial metadata, which should ideally be consistent.